clients/drcachesim/tests/trace_filter_unit_tests.cpp

```diff
@@ -0,0 +1,132 @@
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/* Unit tests for trace_filter. */
+
+#include "analyzer.h"
+#include "dr_api.h"
+#include "droption.h"
+#include "tools/basic_counts.h"
+#include "tools/trace_filter.h"
+
+#define FATAL_ERROR(msg, ...)                               \
+    do {                                                    \
+        fprintf(stderr, "ERROR: " msg "\n", ##__VA_ARGS__); \
+        fflush(stderr);                                     \
+        exit(1);                                            \
+    } while (0)
+
+#define CHECK(cond, msg, ...)             \
+    do {                                  \
+        if (!(cond)) {                    \
+            fprintf(stderr, "%s\n", msg); \
+            return false;                 \
+        }                                 \
+    } while (0)
+
+static droption_t<std::string>
+    op_trace_dir(DROPTION_SCOPE_FRONTEND, "trace_dir", "",
+                 "[Required] Trace input directory",
+                 "Specifies the directory containing the trace files to be filtered.");
+
+static droption_t<std::string> op_tmp_output_dir(
+    DROPTION_SCOPE_FRONTEND, "tmp_output_dir", "",
+    "[Required] Output directory for the filtered trace",
+    "Specifies the directory where the filtered trace will be written.");
+
+static bool
+local_create_dir(const char *dir)
+{
+    if (dr_directory_exists(dir))
+        return true;
+    return dr_create_dir(dir);
+}
+
+basic_counts_t::counters_t
+get_basic_counts(const std::string &trace_dir)
+{
+    analysis_tool_t *basic_counts_tool = new basic_counts_t(/*verbose=*/0);
+    std::vector<analysis_tool_t *> tools;
+    tools.push_back(basic_counts_tool);
+    analyzer_t analyzer(trace_dir, &tools[0], (int)tools.size());
+    if (!analyzer) {
+        FATAL_ERROR("failed to initialize analyzer: %s",
+                    analyzer.get_error_string().c_str());
+    }
+    if (!analyzer.run()) {
+        FATAL_ERROR("failed to run analyzer: %s", analyzer.get_error_string().c_str());
+    }
+    basic_counts_t::counters_t counts =
+        ((basic_counts_t *)basic_counts_tool)->get_total_counts();
+    delete basic_counts_tool;
+    return counts;
+}
+
+bool
+test_nop_filter()
+{
+    std::string output_dir = op_tmp_output_dir.get_value() + "/nop_filter";
+    if (!local_create_dir(output_dir.c_str())) {
+        FATAL_ERROR("Failed to create filtered trace output dir %s", output_dir.c_str());
+    }
+    {
+        auto trace_filter = std::unique_ptr<trace_filter_t>(
+            new trace_filter_t(op_trace_dir.get_value(), output_dir));
```

Comment:
There is no need for this to be heap-allocated: a local var suffices without any need for a unique_ptr.

Reply:
Done.